vllm/timings/timer.py

```diff
@@ -0,0 +1,81 @@
+import logging
+import time
+from contextlib import contextmanager
+from threading import RLock
+
+import numpy
+
+__all__ = ["Timer"]
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class Timer:
+    """Timer to log timings during inference requests. Should be used through 
+        timings.utils.get_singleton_manager()
+    """
+    _instance = None
+
+    def __init__(self, avg_after_iterations: int, enable_logging: bool):
+        """
+        :param avg_after_iterations: the number of iterations that have to occur
+            before the time measurements are averaged and displayed to the user
+        :param enable_logging: whether or not time logging is enabled
+        """
+        self.enable_logging = enable_logging
+        self.measurements = {}
+        self.avg_after_iterations = avg_after_iterations
+        self.current_iter = {}
+        self.lock = RLock()
+
+    @contextmanager
+    def time(self, func_name: str):
+        """
+        Time the given function, add the time to the dictionary of lists 
+        tracking each function and calculate the average, if 
+        self.avg_after_iterations number of iterations have passed.
+
+        :param func_name: name of the function that will be used to log the 
+            measurements
+        """
+        start = time.time()
+        try:
+            yield
+        finally:
+            end = time.time()
+            with self.lock:
+                self._update(func_name, end - start)
+                self._compute_average(func_name)
+
+    def _update(self, func_name: str, time_elapased: float):
+        """
+        Update the dictionary of measurements and counter for each function call
+
+        :param func_name: name of the function that will be used to log the 
+            measurements
+        :param time_elapsed: time taken for function execution  
+        """
+        if self.current_iter.get(func_name) is None:
+            self.current_iter[func_name] = 1
+        else:
+            self.current_iter[func_name] += 1
+
+        if self.measurements.get(func_name) is None:
+            self.measurements[func_name] = [time_elapased]
+        else:
+            self.measurements[func_name].append(time_elapased)
+
+    def _compute_average(self, func_name: str):
```

Comment:
Currently `measurements` can grow to unbounded length. Since vllm is a long running process, this could get us into trouble. 

Additionally, I think we actually care about the average time spent over the last cycle as opposed to time spent overall. As a result, I think it would make sense to clear out all the old measurements each time we compute the average

Reply:
Updated such that old measurements are cleared out every time the new average is calculated.